### PR TITLE
util/types, column, expressions: use types.Convert to replace Col.Cas…

### PR DESCRIPTION
**Pull request by coocood**
…tValue

The two methods has similar functionality and has complex logic, after this change,
the code base would be easier to maintain.

For `CAST` function, the behaviour is different than the general purpose `types.Convert`,
so I implemented a dedicated function for it.

### CODE REVIEW

File: util/types/convert.go [hunk 1]
```diff
@@ -188,6 +405,7 @@ func StrToInt(str string) (int64, error) {
 	if len(str) == 0 {
 		return 0, nil
 	}
+	str = strings.Trim(str, " \t\r\n")
```
Comment (qiuyesuifeng) [inline util/types/convert.go]: extract to a new function? and do trim before check len(str)?

File: util/types/convert.go [hunk 2]
```diff
@@ -39,6 +41,232 @@ func truncateStr(str string, flen int) string {
...
+}
+
+func convertFloatToInt(val float64, lowerBound, upperBound int64, tp byte) (converted int64, err error) {
+	val = RoundFloat(val)
+	if val < math.MinInt64 || int64(val) < lowerBound {
+		converted = lowerBound
+		err = Overflow(val, tp)
+		return
+	}
+	if val > math.MaxInt64 || int64(val) > upperBound {
```
Comment (qiuyesuifeng) [inline util/types/convert.go]: here can only check `int64(val) > upperBound`?

File: util/types/convert.go [hunk 3]
```diff
@@ -39,6 +41,232 @@ func truncateStr(str string, flen int) string {
...
+	case float64:
+		return convertFloatToInt(float64(v), lowerBound, upperBound, tp)
+	case string:
+		fval, err := StrToFloat(v)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		return convertFloatToInt(fval, lowerBound, upperBound, tp)
+	case mysql.Decimal:
+		fval, _ := v.Float64()
```
Comment (qiuyesuifeng) [inline util/types/convert.go]: can use IntParts() for check?

Comment (coocood) [inline util/types/convert.go]: Decimal value may larger than MaxInt64

**Comment by siddontang** [comment]
LGTM

File: util/types/convert.go [hunk 4]
```diff
@@ -39,6 +41,232 @@ func truncateStr(str string, flen int) string {
...
+}
+
+// typeError returns error for invalid value type.
+func typeError(v interface{}, target *FieldType) error {
+	return errors.Errorf("cannot use %v (type %T) in assignment to, or comparison with, column type %s)",
+		v, v, target.String())
+}
+
+// Cast casts val to certain types and does not return error.
+func Cast(val interface{}, target *FieldType) (v interface{}) { //NTYPE
```
Comment (qiuyesuifeng) [inline util/types/convert.go]: what is `NTYPE` in this comment line?

Comment (cznic) [inline util/types/convert.go]: This comment annotates things which had to be updated to handle newly added QL types.

Comment (qiuyesuifeng) [inline util/types/convert.go]: gotcha :)

**Pull request merged**